_posts/2024-08-01-pip-freeze.markdown

```diff
@@ -0,0 +1,119 @@
+</center>
+
+# Key Takeaways
+**Version Control and Backups**: Always keep versions and backups of previous images. Even if the images are created by code, unexpected changes in dependencies can lead to failures. Below is an example of our AWS ECR, and its attached lifecycle policy.
+```hcl
+resource "aws_ecr_repository" "this" {
+  name = local.ecr_repository_name
+  image_scanning_configuration {
+    scan_on_push = true
+  }
+  image_tag_mutability = "IMMUTABLE"
+}
+
+data "aws_ecr_lifecycle_policy_document" "this" {
+  rule {
+    priority    = 1
+    description = "Only keep three (3) images for application repositories"
+    selection {
+      tag_status   = "any"
+      count_type   = "imageCountMoreThan"
+      count_number = 3
+    }
+    action {
+      type = "expire"
+    }
+  }
+}
+
+resource "aws_ecr_lifecycle_policy" "this" {
+  repository = aws_ecr_repository.this.name
+  policy     = data.aws_ecr_lifecycle_policy_document.this.json
+}
+```
+
+**Freezing Package Requirements**: Regularly use pip freeze to capture the exact package versions in use. This practice ensures that your environment remains consistent and prevents unexpected issues due to automatic updates.
+```diff
++ numpy==1.26.4
+- pandas
++ pandas==2.2.2
+- requests
++ requests==2.32.3
+- s3fs
++ s3fs==2024.6.1
+- watchtower
++ watchtower==3.2.0
+- sqlalchemy
++ SQLAlchemy==2.0.31
+- pyodbc
++ pyodbc==5.1.0
+- psycopg2-binary
++ psycopg2-binary==2.9.9
+- pytz
++ pytz==2024.1
+```
+
+**Comprehensive Snapshots**: Using pip freeze provides a complete snapshot of all installed packages and their versions at the time of installation. This snapshot is incredibly valuable for recreating the exact environment later, ensuring consistency across different stages of development and deployment.
+
+**Manual Cleanup**: Over time, your project might evolve, and certain packages may no longer be needed. However, pip freeze will still list these packages unless manually removed. It’s essential to keep track of your dependencies and clean up the ones you no longer use to avoid unnecessary clutter.
+
+**Consider Using pipreqs**: For more refined control over your dependencies, consider using pipreqs. This tool generates a requirements.txt file based on your imports, ensuring you only include the necessary packages. It’s a great way to streamline your dependency management and avoid bloating your project with unused libraries.
+
+# Wrapping Up
+
+pip freeze is a vital tool for data engineers and analysts. It helps maintain a stable and predictable environment, allowing you to focus on building robust data solutions without worrying about unexpected dependency issues. By adopting these best practices, you can ensure that your projects remain consistent, reliable, and easier to manage. So, don't just burn through your data—freeze your requirements and safeguard your workflows against the chaos of untracked dependencies.
```

Comment:
maybe just add some kind of Call To Action... ie. get in touch at some email address or a link to our website etc